design/src/rfcs/rfc0023_sdk_credential_cache_type_safety.md

```diff
@@ -0,0 +1,328 @@
+    Custom(Box<dyn SomeCacheTrait>),
+
+    NoCaching,
+}
+pub struct CredentialsCache {
+    inner: Inner,
+}
+
+impl CredentialsCache {
+    // Methods prefixed with `default_` just use the default cache settings
+    pub fn default_lazy() -> Self { /* ... */ }
+    pub fn default_eager() -> Self { /* ... */ }
+
+    // Unprefixed methods return a builder that can take customizations
+    pub fn lazy() -> LazyBuilder { /* ... */ }
+    pub fn eager() -> EagerBuilder { /* ... */ }
+
+    // Later, when custom implementations are supported
+    pub fn custom(cache_impl: Box<dyn SomeCacheTrait>) -> Self { /* ... */ }
+
+    pub(crate) fn create_cache(
+        self,
+        provider: Box<dyn ProvideCredentials>,
+        sleep_impl: Arc<dyn AsyncSleep>
+    ) -> SharedCredentialsProvider {
+        SharedCredentialsProvider::new(
+            match self {
+                Self::Lazy(inner) => LazyCachingCredentialsProvider::new(provider, settings.time, /* ... */),
+                Self::Eager(_inner) => unimplemented!(),
+                Self::Custom(_custom) => unimplemented!(),
+                Self::NoCaching => unimplemented!(),
+            }
+        )
+    }
+}
+```
+
+Using a struct over a trait prevents custom caching implementations, but if customization is desired,
+a `Custom` variant could be added to the inner enum that has its own trait that customers implement.
+
+The `SharedCredentialsProvider` needs to be updated to take a cache implementation in addition to
+the `impl ProvideCredentials + 'static`. A sealed trait could be added to facilitate this.
+
+Customers that don't care about credential caching can configure credential providers
+without needing to think about it:
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials_provider(ImdsCredentialsProvider::builder().build())
+    .load()
+    .await;
+```
+
+However, if they want to customize the caching, they can do so without modifying
+the credentials provider at all (in case they want to use the default):
+
+```rust
+let sdk_config = aws_config::from_env()
+    .credentials_cache(CredentialsCache::default_eager())
+    .load()
+    .await;
+```
+
+The `credentials_cache` will default to `CredentialsCache::default_lazy()` if not provided.
+
+Changes Checklist
+-----------------
+
+- [ ] Implement `CredentialsCache` with its `Lazy` variant and builder
+- [ ] Add `credentials_cache` method to `ConfigLoader`
+- [ ] Refactor `ConfigLoader` to take `CredentialsCache` instead of `impl ProvideCredentials + 'static`
+- [ ] Refactor `SharedCredentialsProvider` to take a cache implementation in addition to an `impl ProvideCredentials + 'static`
+- [ ] Rename `LazyCachingCredentialsProvider` -> `LazyCredentialsCache`
+- [ ] Refactor the SDK `Config` code generator to be consistent with `ConfigLoader`
+- [ ] Write changelog upgrade instructions
+- [ ] Fix examples (if there are any for configuring caching)
+
```

Comment:
Related to [this](https://github.com/awslabs/smithy-rs/pull/1842#discussion_r1013228538) comment, we may also need
- [ ] Remove a field [cache](https://github.com/awslabs/smithy-rs/blob/fba88df3eeb6c5ea8be2ea2df57232366c016b37/aws/rust-runtime/aws-config/src/sts/assume_role.rs#L47) from `AssumeRoleProvider`